consensus/src/chained_bft/network.rs

```diff
@@ -274,142 +252,357 @@ impl NetworkSender {
-            QueueStyle::LIFO,
-            NonZeroUsize::new(1).unwrap(),
-            Some(&counters::BLOCK_RETRIEVAL_CHANNEL_MSGS),
-        );
-        let (sync_info_tx, sync_info_rx) = libra_channel::new(
-            QueueStyle::LIFO,
-            NonZeroUsize::new(1).unwrap(),
-            Some(&counters::SYNC_INFO_CHANNEL_MSGS),
-        );
-        let (epoch_change_tx, epoch_change_rx) = libra_channel::new(
-            QueueStyle::LIFO,
-            NonZeroUsize::new(1).unwrap(),
-            Some(&counters::EPOCH_CHANGE_CHANNEL_MSGS),
-        );
-        let (different_epoch_tx, different_epoch_rx) =
-            libra_channel::new(QueueStyle::LIFO, NonZeroUsize::new(1).unwrap(), None);
-        let (epoch_retrieval_tx, epoch_retrieval_rx) =
-            libra_channel::new(QueueStyle::LIFO, NonZeroUsize::new(1).unwrap(), None);
-        let network_events = network_events.map_err(Into::<anyhow::Error>::into);
-        let all_events = Box::new(select(network_events, self_receiver));
-        (
-            NetworkTask {
-                epoch_info,
-                proposal_tx,
-                vote_tx,
-                block_request_tx,
-                sync_info_tx,
-                epoch_change_tx,
-                different_epoch_tx,
-                epoch_retrieval_tx,
-                all_events,
-            },
-            NetworkReceivers {
-                proposals: proposal_rx,
-                votes: vote_rx,
-                block_retrieval: block_request_rx,
-                sync_info_msgs: sync_info_rx,
-                epoch_change: epoch_change_rx,
-                different_epoch: different_epoch_rx,
-                epoch_retrieval: epoch_retrieval_rx,
-            },
-        )
+        network_data_request_receiver: mpsc::UnboundedReceiver<ConsensusDataRequest>,
+        network_data_sender: mpsc::UnboundedSender<(AccountAddress, FromNetworkMsg<T>)>,
+    ) -> NetworkTask<T> {
+        NetworkTask {
+            incoming: Default::default(),
+            round_robin_queue: Default::default(),
+            network_data_request_receiver,
+            network_data_sender,
+        }
     }
 
-    fn epoch(&self) -> u64 {
-        self.epoch_info.read().unwrap().epoch
-    }
+    pub async fn start(
+        mut self,
+        all_events: Box<dyn Stream<Item = anyhow::Result<Event<ConsensusMsg>>> + Send + Unpin>,
+    ) {
+        // If consensus asked for data, and we had none to give,
+        // this flag is set,
+        // so that when new data comes in over the network,
+        // it is immediately sent over to consensus without first requiring another data request.
+        let mut consensus_needs_data = false;
+
+        let mut all_events = all_events.fuse();
+        loop {
+            select! {
+                message = all_events.next() => {
+                    match message {
+                        Some(Ok(msg)) => self.handle_msg(msg).await,
+                        _ => continue,
```

Comment:
I'm actually not sure, could be that the stream has ended? It might actually always be `Ok` here. 

There is a `complete` handler below, which I believe is called when the select is "completed". 

In terms of exiting, my understanding is that when the `ConsensusProvider` drops the runtime, all tasks are cancelled. This happens when `stop` is called.

Reply:
Switched to using `select_next_some` on this one...